feat: capture metadata

ShaswatPrabhat opened the pull request:
Changes for [isssue](https://github.com/callstack/reassure/issues/60)

mdjastrzebski commented:
@ShaswatPrabhat good work! pls add screenshots of current console & markdown output

ShaswatPrabhat commented:
> @ShaswatPrabhat good work! pls add screenshots of current console & markdown output
Console:
<img width="805" alt="image" src="https://user-images.githubusercontent.com/11253143/193443007-0e2f66b3-411a-4bd3-9945-48a079ddcf78.png">

JSON:
<img width="784" alt="image" src="https://user-images.githubusercontent.com/11253143/193443039-7484cd99-d8c3-429e-89ac-598dc6bfd20a.png">

Markdown:

<img width="841" alt="image" src="https://user-images.githubusercontent.com/11253143/193443070-a1a21322-0c7b-4105-97bf-b286bb542c03.png">

mdjastrzebski commented on packages/reassure-cli/src/commands/measure.ts:
```diff
@@ -22,6 +24,19 @@ export function run(options: MeasureOptions) {
   const outputFile = options.baseline ? BASELINE_FILE : RESULTS_FILE;
   rmSync(outputFile, { force: true });
 
+  const metadata = {
+    metadata: {
+      branch: options.branch,
+      commitHash: options.commitHash,
+    },
+  };
```
```suggestion
  const header = {
    metadata: {
      branch: options.branch,
      commitHash: options.commitHash,
    },
  };
```

mdjastrzebski commented on packages/reassure-cli/src/commands/measure.ts:
```diff
@@ -22,6 +24,19 @@ export function run(options: MeasureOptions) {
+    },
+  };
+
+  const metaDataString = JSON.stringify(metadata) + '\n';
+
+  writeFileSync(outputFile, metaDataString);
```
```suggestion
  writeFileSync(outputFile, JSON.stringify(header) + '\n');
```

mdjastrzebski commented on packages/reassure-cli/src/commands/measure.ts:
```diff
@@ -22,6 +24,19 @@ export function run(options: MeasureOptions) {
+
+  writeFileSync(outputFile, metaDataString);
+
+  console.log('');
```
Is there a reason for adding this console empty line here?

ShaswatPrabhat replied:
Sorry I will remove this right away

ShaswatPrabhat replied:
Have addressed this in latest commits.

mdjastrzebski commented on packages/reassure-cli/src/commands/measure.ts:
```diff
@@ -85,6 +100,16 @@ export const command: CommandModule<{}, MeasureOptions> = {
+      .option('branch', {
+        type: 'string',
+        default: '--not-specified--',
+        describe: 'Metadata for branch name to be included in the report',
```
```suggestion
        describe: 'Source control branch for which the measurement is taken',
```

mdjastrzebski commented on packages/reassure-cli/src/commands/measure.ts:
```diff
@@ -85,6 +100,16 @@ export const command: CommandModule<{}, MeasureOptions> = {
+      })
+      .option('branch', {
+        type: 'string',
+        default: '--not-specified--',
```
I think that there should be no default value atm. In the subsequent PR we will try to get it from `git`.
```suggestion
```

mdjastrzebski commented on packages/reassure-cli/src/commands/measure.ts:
```diff
@@ -85,6 +100,16 @@ export const command: CommandModule<{}, MeasureOptions> = {
+      .option('commitHash', {
+        type: 'string',
+        default: '--not-specified--',
+        describe: 'Commit hash, will be included in the report',
```
```suggestion
        describe: 'Git commit hash of current code, it will be included in the report',
```

mdjastrzebski commented on packages/reassure-compare/src/compare.ts:
```diff
@@ -56,33 +65,32 @@ export async function compare({
   }
 
-  const current = await loadFile(currentFile);
+  const currentMetadata = await loadMetadataFromFile(currentFile);
```
Maybe we could tweak the original method to return structured data:

```suggestion
  const { metadata, results } = await loadFile(currentFile);
```

ShaswatPrabhat replied:
The reason I had made a separate method for these was because I saw a TODO around name tag validation.

I thought if we could have separate methods for each validation then it would make more sense so that I could then refactor out `readFileAndParseDataToFormat` to make some validations based on tagnames

ShaswatPrabhat replied:
But I get your point, I will re-add the older method and make it load both performance results and metadata

ShaswatPrabhat replied:
Have addressed this in latest commits.

mdjastrzebski commented on packages/reassure-compare/src/compare.ts:
```diff
@@ -100,13 +108,40 @@ async function loadFile(path: string): Promise<PerformanceRecord> {
+    logError(msg);
+    throw new Error(msg);
+  }
+  if (entries.length == 0) {
```
I think we should allow for missing metadata, as this will be the case when running baseline calculation after interoducing metadata features, the baseline will still use the previous Reassure and will not insert that entry.

mdjastrzebski commented on packages/reassure-compare/src/compare.ts:
```diff
@@ -100,13 +108,40 @@ async function loadFile(path: string): Promise<PerformanceRecord> {
  * Compare results between baseline and current entries and categorize.
  */
-function compareResults(currentEntries: PerformanceRecord, baselineEntries: PerformanceRecord | null): CompareResult {
+function compareResults(
+  currentEntries: PerformanceRecord,
+  baselineEntries: PerformanceRecord | null,
+  currentMetadata: MeasurementMetadata,
+  baselineMetadata: MeasurementMetadata | null
+): CompareResult {
```
```suggestion
function compareResults(
  current: MeasurementResults,
  baseline: MeasurementResults | null,
): CompareResult {
```

ShaswatPrabhat replied:
Have addressed this in latest commits.

mdjastrzebski commented on packages/reassure-compare/src/compare.ts:
```diff
@@ -144,6 +181,7 @@ function compareResults(currentEntries: PerformanceRecord, baselineEntries: Perf
     countChanged,
     added,
     removed,
+    metadata,
```
Let's move `metadata` as the first prop of the return, before `errors`

mdjastrzebski commented on packages/reassure-compare/src/output/console.ts:
```diff
@@ -24,6 +27,11 @@ export function printToConsole(data: CompareResult) {
   console.log('');
 }
 
+function printMetadata(entry: Metadata) {
+  console.log(
+    ` - baseline: ${entry.baseline?.branch} :: ${entry.baseline?.commitHash} | current: ${entry.current?.branch} :: ${entry.current?.commitHash}`
+  );
+}
```
I would rather not assume that we have either piece of metadata, and do something like below:

```suggestion
function buildMetadata(name: string, metadata?: MeasurementMetadata) {
  if (metadata.branch && metadata.commitHash) {
     return `**${name}**: \`${metadata.branch}\` (\`${metadata.commitHash}\`)`;
  }

  if (metadata.branch) {
    return `**\`${name}**\`: ${\`metadata.branch\`}`;
  }

  if (metadata.branch) {
    return `**${name}**: \`${metadata.commitHash}\``;
  }

  return `**${name}**: missing metadata`;
}
```

mdjastrzebski commented on packages/reassure-compare/src/output/markdown.ts:
```diff
@@ -39,9 +46,30 @@ async function writeToFile(filePath: string, content: string) {
 }
 
+function buildMetadataMarkdown(name: string, metadata: MeasurementMetadata | null) {
+  if (metadata) {
+    if (metadata.branch && metadata.commitHash) {
+      return `**${name}**: \`${metadata.branch}\` (\`${metadata.commitHash}\`)`;
+    }
+
+    if (metadata.branch) {
+      return `**${name}**: \`${metadata.branch}\``;
+    }
+
+    if (metadata.commitHash) {
+      return `**${name}**: \`${metadata.commitHash}\``;
+    }
+  }
```
nit: Using optional chaining should allow for flattening of `if`s
```suggestion
    if (metadata?.branch && metadata?.commitHash) {
      return `**${name}**: \`${metadata.branch}\` (\`${metadata.commitHash}\`)`;
    }

    if (metadata?.branch) {
      return `**${name}**: \`${metadata.branch}\``;
    }

    if (metadata?.commitHash) {
      return `**${name}**: \`${metadata.commitHash}\``;
    }
```

mdjastrzebski commented on packages/reassure-compare/src/types.ts:
```diff
@@ -1,6 +1,23 @@
+  commitHash: string;
+};
+
+export interface MetadataEntry {
+  /** Metadata */
+  metadata: MeasurementMetadata;
+}
```
```suggestion
export interface PerformanceHeader {
  /** Metadata */
  metadata: MeasurementMetadata;
}
```

ShaswatPrabhat replied:
Have addressed this in latest commits.

mdjastrzebski commented on packages/reassure-compare/src/types.ts:
```diff
@@ -57,6 +74,10 @@ export interface RemovedEntry {
   baseline: PerformanceEntry;
 }
 
+export interface Metadata {
```
Since this type is for compare output lets add `Compare` to the name
```suggestion
export interface CompareMetadata {
```

mdjastrzebski commented on packages/reassure-compare/src/utils/readFileAndParseDataToFormat.ts:
```diff
@@ -0,0 +1,15 @@
+
+import fs from 'fs/promises';
+
+export async function readFileAndParseDataToFormat(path: string, keyToFilterWith: string) {
```
I think this function got a bit too generic with the `keyToFilterWith` prop. I would be happy to trade abstraction here to a small amount of `if` statements in the original `loadFile` function

mdjastrzebski reviewed (changes requested):
Awesome work @ShaswatPrabhat !

I've left few minor comments to improve code readability/structure.

Please update also README.me section named `### Write performance testing script` by passing these params from git.

Also update our own `reassure-tests.sh` in `examples/native` and other examples to capture branch name and commit hash from git.
```
yarn reassure --branch `git branch --show-current` --commitHash `git rev-parse HEAD`
```

`git branch --show-current`
`git rev-parse HEAD`

mdjastrzebski commented on packages/reassure-compare/src/output/console.ts:
```diff
@@ -24,6 +27,25 @@ export function printToConsole(data: CompareResult) {
+  }
+
+  if (metadata.branch) {
+    console.log(` - \`${name}\`: ${\`metadata.branch\`}`);
```
```suggestion
    console.log(` - ${name}: ${metadata.branch}`);
```

mdjastrzebski commented on packages/reassure-compare/src/output/console.ts:
```diff
@@ -24,6 +27,25 @@ export function printToConsole(data: CompareResult) {
 
+function printMetadata(name: string, metadata?: MeasurementMetadata) {
+  if (metadata.branch && metadata.commitHash) {
+     console.log(`- ${name}: \`${metadata.branch}\` (\`${metadata.commitHash}\`)`);
```
Seems like you have extra back quotes here.
```suggestion
     console.log(`- ${name}: ${metadata.branch} (${metadata.commitHash})`);
```

mdjastrzebski commented on packages/reassure-compare/src/output/console.ts:
```diff
@@ -24,6 +27,25 @@ export function printToConsole(data: CompareResult) {
+  }
+
+  if (metadata.branch) {
+    console.log(`- ${name}: \`${metadata.commitHash}\``);
```
```suggestion
    console.log(`- ${name}: ${metadata.commitHash}`);
```

ShaswatPrabhat commented:
> Awesome work @ShaswatPrabhat !
> 
> I've left few minor comments to improve code readability/structure.
> 
> Please update also README.me section named `### Write performance testing script` by passing these params from git.
> 
> Also update our own `reassure-tests.sh` in `examples/native` and other examples to capture branch name and commit hash from git.
> 
> ```
> yarn reassure --branch `git branch --show-current` --commitHash `git rev-parse HEAD`
> ```
> 
> `git branch --show-current` `git rev-parse HEAD`

Have addressed all comments and README and native script changes @mdjastrzebski.
Please help review and merge

mdjastrzebski commented on .changeset/eleven-yaks-love.md:
```diff
@@ -0,0 +1,6 @@
+'@callstack/reassure-compare': minor
+---
+
+The changes integrate metatda information in the performanc file
```
```suggestion
Include codebase metadata (branch, commit hash) in the performance report and measurement file
```

mdjastrzebski commented on packages/reassure-compare/src/types.ts:
```diff
@@ -1,6 +1,23 @@
 /**
  * Entry in the performance results file.
  */
+
+/**
+ * Metadata first line of file
+ */
+
+export type MeasurementMetadata = {
+  /** Name of branch */
+  branch: string;
+  /** Commit Hash */
+  commitHash: string;
+};
+
+export interface PerformanceHeader {
+  /** Metadata */
+  metadata: MeasurementMetadata;
+}
+
```
```suggestion
/** Metadata information for measurements */
export type MeasurementMetadata = {
  branch: string;
  commitHash: string;
};

/** Header of performance results file. */
export interface PerformanceHeader {
  metadata: MeasurementMetadata;
}

/** Entry in the performance results file. */
```

mdjastrzebski reviewed (approved):
Awesome stuff @ShaswatPrabhat 🚀 

I've made some final tweaks, and let's merge it!
